miniapps/blocksolvers/div_free_solver.hpp

```diff
@@ -0,0 +1,252 @@
+{
+   const SparseMatrix& agg_hdivdof_;
+   const SparseMatrix& agg_l2dof_;
+   OperatorPtr coarse_l2_projector_;
+
+   Array<int> offsets_;
+   mutable Array<int> offsets_loc_;
+   mutable Array<int> hdivdofs_loc_;
+   mutable Array<int> l2dofs_loc_;
+   Array<OperatorPtr> solvers_loc_;
+public:
+   SchwarzSmoother(const HypreParMatrix& M,
+                   const HypreParMatrix& B,
+                   const SparseMatrix& agg_hdivdof,
+                   const SparseMatrix& agg_l2dof,
+                   const HypreParMatrix& P_l2,
+                   const HypreParMatrix& Q_l2);
+   virtual void Mult(const Vector & x, Vector & y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Relaxation on the kernel of divergence
+class KernelSmoother : public Solver
+{
+   Array<int> offsets_;
+   OperatorPtr blk_kernel_map_;
+   OperatorPtr kernel_system_;
+   OperatorPtr kernel_smoother_;
+public:
+   KernelSmoother(const BlockOperator& op, const HypreParMatrix& kernel_map_);
+   virtual void Mult(const Vector & x, Vector & y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Solver S such that I - A * S = (I - A * S1) * (I - A * S0).
+/// A, S0, S1 are assumed to be symmetric.
+class ProductSolver : public Solver
+{
+   OperatorPtr op_;
+   Array<OperatorPtr> solvers_;
+   void Mult(int i, int j, const Vector & x, Vector & y) const;
+public:
+   ProductSolver(Operator* A, Operator* S0, Operator* S1,
+                 bool ownA, bool ownS0, bool ownS1)
+      : Solver(A->NumRows()), op_(A, ownA), solvers_(2)
+   {
+      solvers_[0].Reset(S0, ownS0);
+      solvers_[1].Reset(S1, ownS1);
+   }
+   virtual void Mult(const Vector & x, Vector & y) const { Mult(0, 1, x, y); }
+   virtual void MultTranspose(const Vector & x, Vector & y) const { Mult(1, 0, x, y); }
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Divergence free solver, for more details see Appendix F.3 in
+/// Vassilevski, Multilevel Block Factorization Preconditioners, 2008
+class DivFreeSolver : public DarcySolver
+{
+   const DFSData& data_;
+   OperatorPtr BT_;
+   BBTSolver BBT_solver_;
+   OperatorPtr CTMC_;
+   Array<Array<int>> ops_offsets_;
+   Array<OperatorPtr> ops_;
+   Array<OperatorPtr> blk_Ps_;
+   Array<OperatorPtr> smoothers_;
+   OperatorPtr prec_;
+   OperatorPtr solver_;
+
+   void SolveParticular(const Vector& rhs, Vector& sol) const;
+   void SolveDivFree(const Vector& rhs, Vector& sol) const;
+   void SolvePotential(const Vector &rhs, Vector& sol) const;
+public:
+   DivFreeSolver(const HypreParMatrix& M, const HypreParMatrix &B,
+                 const DFSData& data);
+   virtual void Mult(const Vector & x, Vector & y) const;
+   virtual void SetOperator(const Operator &op) { }
+   virtual int GetNumIterations() const
+   {
+      return solver_.As<IterativeSolver>()->GetNumIterations();
+   }
+};
+
+/// Multigrid class that uses HypreSmoother if smoothers are not given
+class AbstractMultigrid : public Solver
```

Comment:
Can you use the `Multigrid` class that was recently added to MFEM? It should do almost exactly what you have here.

Reply:
(Pull request #1582 may be relevant here.)

Reply:
That's a good point, I thought about it. My only concern is, I do not have a `FiniteElementSpaceHierarchy`. Can we either make it optional in `Multigrid` or have a parent class that do not require information of finite element spaces?

Reply:
I actually think the right answer is to make a more abstract parent class for `Multigrid`, but that makes this PR more high-impact. If you think you can make that work with relatively few lines of code I would go that direction but if not maybe what you have now is okay.